### ISSUE TITLE
Make 3p iframe subdomains unique but deterministic.

### ISSUE THREAD

**Issue by mgriego** [ampproject/amphtml#27091]
## Summary

Make 3p iframe subdomains unique but deterministic.

## Motivation

Currently, the 3p iframes that are used for ads and other sandboxed features utilize fully random hostnames on each page load.  All 3p iframes on the current page use the same origin so that, among themselves, they are same-origin but cross-origin to the parent document.  However, on each page load, the hostname/origin that is used to create the 3p iframes is generated based on a new random value.  So, multiple accesses to the same page will never utilize the same origin for the 3p iframes.

I suspect that the above was done in order to protect from information leakage between unaffiliated AMP requests (ie not leaking information between different pages).  This is a very valid privacy and security issue that should remain in place, however utilizing a fully-random hostname causes issues for some implementation, such as an ad network that might try to cache information in localStorage, for instance.  The use of the random origins means that a new localStorage instance will be used on each and every page load.

For the above reasons, I propose changing the 3p hostname generation logic to move away from `d-<RANDOMNUMBER>.ampproject.net` to using something like `d-<BASE32_ENCODED_SHA256_SOURCE_HOSTNAME_HASH>.ampproject.net`.  (Base32/SHA256 was chosen simply because that's the fallback mechanism used by the cache domain generation algorithm already.)  This would maintain the cross-origin requirement with the top-level document and also maintain the origin separation between unaffiliated AMP documents (ie between pages coming from `example1.com` vs `example2.com`), helping to secure against information leakage, however it would also mean that the 3p iframes use repeating, deterministic
origins allowing things like localStorage to work as intended.
## Launch tracker

TBD

/cc @ampproject/wg-ads

**Comment by jridgewell** [comment]
@ampproject/wg-approvers @ampproject/wg-ads 

> however utilizing a fully-random hostname causes issues for some implementation, such as an ad network that might try to cache information in localStorage, for instance. The use of the random origins means that a new localStorage instance will be used on each and every page load.

This is actually the reason we made it random. 3p frames aren't for storing PII.

**Comment by mgriego** [comment]
Who said anything about PII?  Sure, that's a potential use case, but there are other very valid use cases for localStorage within 3p iframes.  You've unilaterally made a judgement call that potentially negatively affects third parties running within AMP.....

**Comment by cramforce** [comment]
This is worth considering. The current setup is for security threat model which *I think* the proposed model addresses. Over to @lannka who can also run it by security.

**Comment by powerivq** [comment]
Don't know enough about the threat model to tell if it's sufficient. Question: Since safeframe uses a static hostname, would that threat model apply equally to safeframe as well?

**Comment by cramforce** [comment]
@powerivq Same host name allows

- Different ads/3p embeds to attack each other
- Different ads/3p embeds embedded into different documents in the same frame hierarchy (happens in the AMP Viewer situation) to attack each other.

**Comment by powerivq** [comment]
@cramforce Thanks. I see the current implementation uses the same random host name for all `<amp-embed` on the same page. Is `amp-embed`s on the same ampdoc accessing each other considered okay?

I am raising safeframe because it uses a globally static host name (www.googlesyndication.com) and runs untrusted code. In theory I believe they could also attach each other? We might also want to fix them.

**Comment by lannka** [comment]
My understanding that we want to prevent vendors attacking each others. I don't think same vendor across sites is an issue as they can always find a way to communicate to each other, for example having another nested iframe.

So the proposal here should be changed to use vendor name in the host name.

**Comment by powerivq** [comment]
Just to confirm. It will be vendor name only, without regards to origin hostname?

**Comment by lannka** [comment]
Ye, that's what I was thinking. We might even let vendors to provide their own domain to host the frame. @cramforce do you see immediate no-go?

**Comment by mgriego** [comment]
I would argue that the origin hostname is important.  So, for instance, perhaps a hash of the vendor and the origin hostname?

One thing I don't understand about the vendor comments, though, is aren't all 3p iframes within the same page utilizing the same hostname currently?  So different vendors running within different 3p iframes are actually, at least right now, same-origin to each other, are they not?

**Comment by cramforce** [comment]
We actually allow pages to override the hostname. The vendor-hostname topic is completely different. We could do that. But the whole point of the 3P system is that the inner-frame is controlled by us.

Let's focus on this proposal.

**Comment by cramforce** [comment]
My last message had not seen the whole conversation.

- Lots of vendors run untrusted code, so multiple instances are not trusting each other. A change that would remove this scenario from the thread model would be much harder to do.
- We do use one hostname per page load as a trade-off of security and performance.
- I think that literally doing `hash(hostname)` is fine. This would be a nice performance win.

**Comment by lannka** [comment]
> One thing I don't understand about the vendor comments, though, is aren't all 3p iframes within the same page utilizing the same hostname currently? So different vendors running within different 3p iframes are actually, at least right now, same-origin to each other, are they not?

You're actually right. I read the code again, vendors are not isolated at this moment. Any objection we add vendor into the hash input?  @cramforce 

One thing worth to call out: we might pick the sourceUrl origin as the origin for hashing (as opposed to the CDN domain origin). So that the localStorage can be shared across CDN and pub domain.

```
hash(ampSourceUrlOrigin, vendor)
```

**Comment by cramforce** [comment]
As discussed, offline I don't see a need for vendor isolation (not provided
right now) and I'd like to avoid the drop in cacheability and increase in
site-isolation memory usage.

**Comment by mgriego** [comment]
Agreed on both counts - not increasing the scope to include vendor isolation in this but also hashing based on the source URL instead of the proxy URL.

**Comment by lannka** [comment]
@mgriego feel free to draft a PR with your proposed changes

**Comment by lannka** [comment]
@ampproject/wg-security-privacy for approval

**Comment by honeybadgerdontcare** [comment]
@molnarg for feedback

**Comment by molnarg** [comment]
From security PoV, the only potential issue I see here is hash collision between different sites that allow accessing someone else's data. SHA256 is collision resistant, so we're fine on that front.

Enabling the storage of data by vendors would need to be looked into by privacy folks.

I agree that further segmenting 3P domains into vendor specific ones would be a win for security/privacy, but treating that change separately and weighing it against performance sounds good.

> I am raising safeframe because it uses a globally static host name (www.googlesyndication.com) and runs untrusted code. In theory I believe they could also attach each other? We might also want to fix them.

Yes, this is known issue which is being worked on.

**Comment by powerivq** [comment]
@honeybadgerdontcare only saw you two are on wg-security-privacy. Do you know who else may look at the privacy part? Or does the privacy discussion above suffice?

**Comment by honeybadgerdontcare** [comment]
@powerivq We should have a look at the storage of data by vendors by some privacy folks. Replied to you offline to begin that discussion.

**Comment by Tudmotu** [comment]
Hi @powerivq @honeybadgerdontcare, are there any updates here?
We (Taboola) would love to see this merged, so if there is anything we can do to help push this forward, please let us know! 🙂

**Comment by powerivq** [comment]
Hi @Tudmotu 

We want to launch this. Unfortunately since this domain is owned by Google, we need to go through some internal process and it is taking way too long. I do not have an exact ETA but we are working on it.

**Comment by Tudmotu** [comment]
@powerivq Thanks for the update :)
